extmod/modssl_mbedtls: Add SSLContext methods.

Carglglz opened the pull request:
This commit adds:

1) Methods to SSLContext class that match CPython signature:

	- `SSLContext.load_cert_chain(certfile, keyfile)`
	- `SSLContext.load_verify_locations(cafile=, cadata=)`
	- `SSLContext.get_ciphers()` --> ["CIPHERSUITE"]
	- `SSLContext.set_ciphers(["CIPHERSUITE"])`

2) `sslsocket.cipher()` to get current ciphersuite and protocol
   version.

3) `ssl.MBEDTLS_VERSION` string constant

4) Certificate verification errors info instead of `MBEDTLS_ERR_X509_CERT_VERIFY_FAILED`

5) Tests in `net_inet` and `multi_net`

`SSLContext.load_cert_chain` method allows loading key and cert from disk passing a filepath in `certfile` or `keyfile` options.

`SSLContext.load_verify_locations`'s `cafile` option enables the same functionality for ca files.


This may close #10832 , #9071 and #8915

Carglglz committed:
extmod/modssl_mbedtls: Add SSLContext certificate methods.

This commit adds:

1) Methods to SSLContext class that match CPython signature:

	- `SSLContext.load_cert_chain(certfile, keyfile)`
	- `SSLContext.load_verify_locations(cafile=, cadata=)`
	- `SSLContext.get_ciphers()` --> ["CIPHERSUITE"]
	- `SSLContext.set_ciphers(["CIPHERSUITE"])`

2) `sslsocket.cipher()` to get current ciphersuite and protocol
   version.

3) `ssl.MBEDTLS_VERSION` string constant.

4) Certificate verification errors info instead of
   `MBEDTLS_ERR_X509_CERT_VERIFY_FAILED`.

5) Tests in `net_inet` and `multi_net` to test these new methods.

`SSLContext.load_cert_chain` method allows loading key and cert from disk
passing a filepath in `certfile` or `keyfile` options.

`SSLContext.load_verify_locations`'s `cafile` option enables the same
functionality for ca files.

Signed-off-by: Carlos Gil <carlosgilglez@gmail.com>

Carglglz commented:
TODO:
- [ ] `SSLContext.set_ciphers` implementation (see comments in #11897 ) 
- [ ] Buffer size/implementation for cert verify errors.
- [ ] `SSLContext.load_cert_chain(certfile, keyfile=)` MicroPython signature.

Carglglz commented:
https://github.com/micropython/micropython/pull/11897#discussion_r1410070589

@dpgeorge checking `mbedtls` programs, they use 512,  so this may be a sensible default?, 
otherwise if the buffer is too small the error info may be truncated?

```C
mbedtls$ rg vrfy_buf
programs/x509/cert_app.c
331:                char vrfy_buf[512];
335:                mbedtls_x509_crt_verify_info(vrfy_buf, sizeof(vrfy_buf), "  ! ", flags);
337:                mbedtls_printf("%s\n", vrfy_buf);

programs/ssl/ssl_mail_client.c
195:        char vrfy_buf[512];
201:        mbedtls_x509_crt_verify_info(vrfy_buf, sizeof(vrfy_buf), "  ! ", flags);
203:        mbedtls_printf("%s\n", vrfy_buf);

programs/ssl/ssl_client1.c
207:        char vrfy_buf[512];
213:        mbedtls_x509_crt_verify_info(vrfy_buf, sizeof(vrfy_buf), "  ! ", flags);
215:        mbedtls_printf("%s\n", vrfy_buf);

programs/ssl/dtls_client.c
236:        char vrfy_buf[512];
242:        mbedtls_x509_crt_verify_info(vrfy_buf, sizeof(vrfy_buf), "  ! ", flags);
244:        mbedtls_printf("%s\n", vrfy_buf);

programs/ssl/ssl_server2.c
3431:            char vrfy_buf[512];
3434:            x509_crt_verify_info(vrfy_buf, sizeof(vrfy_buf), "  ! ", flags);
3436:            mbedtls_printf("%s\n", vrfy_buf);
3489:        char vrfy_buf[512];
3493:        x509_crt_verify_info(vrfy_buf, sizeof(vrfy_buf), "  ! ", flags);
3494:        mbedtls_printf("%s\n", vrfy_buf);

programs/ssl/ssl_client2.c
2347:        char vrfy_buf[512];
2350:        x509_crt_verify_info(vrfy_buf, sizeof(vrfy_buf),
2353:        mbedtls_printf("%s\n", vrfy_buf);
```

Other than this I will consider this PR done so let me know what you think 👍🏼

dpgeorge commented on tests/multi_net/sslcontext_server_client_ciphers.py:
```diff
@@ -0,0 +1,62 @@
+# Possible cause:
+#   Incompatible ciphersuites + streams.
+#   i.e. error OSError: (-29568, 'MBEDTLS_ERR_SSL_NO_CIPHER_CHOSEN')
+# Action: SKIP for now.
```
Did you test this after all the changes?  It may be working now on 32-bit systems.

Carglglz replied:
I updated the tests and it seems all tests are passing in 32-bit systems too! 🎉

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -236,6 +264,60 @@ STATIC mp_obj_t ssl_context___del__(mp_obj_t self_in) {
+        ssl_context->ciphersuites[len + 1] = 0;
+
+        // Ciphersuite Config
+        if (ssl_context->ciphersuites[0] != 0) {
```
This will always be true, because of the `if (len > 0)` check above, so you can remove this if.

dpgeorge commented:
> checking `mbedtls` programs, they use 512, so this may be a sensible default?,
> otherwise if the buffer is too small the error info may be truncated?

512 bytes is a lot of stack space to allocate.  It's not critical for the error message to fit, even if it's truncated at least it gives a good indication of the problem.

I suggest using 256.  That's a good balance of stack usage and available space for the error.

Otherwise it could be dynamically allocated memory.

dpgeorge committed:
lib/mbedtls_errors: Update error list for latest esp32 mbedtls.

Running `./do-esp32.sh` now generates this esp32_mbedtls_errors.c file,
with IDF v5.0.4.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
esp32: Re-enable custom mbedtls error string tables.

To match other ports that use mbedtls.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
This looks good now.  I have adjusted some of the tests so they pass on bare-metal targets (tested stm32 and esp32).

Thanks @Carglglz for all of your hard work on this, and replying promptly to all of the many code reviews!
